rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+    codebase.
+
+**Defensive**
+
+-   The design of `tf.Module` should optimize for helping avoid common mistakes
+    (e.g. accidentally not tracking variables).
+-   If functionality is required (e.g. calling `super(..).__init__()` in
+    subclasses) we should error early (e.g. immediately after the module
+    constructor returns) to remind users if they forget.
+
+## Detailed Design
+
+### Class structure
+
+We propose a pure Python class `tf.Module` which has a metaclass
+`tf.ModuleMetaclass` to wrap all methods in the modules name scope. This design
+allows for submodules or variables to created in any method (including
+`__init__` and `__call__`). There is more detail on this decision in the
+detailed design below.
+
+```python
+class Module(six.with_metaclass(ModuleMetaclass, tf.checkpoint.Checkpointable)):
+  """Base neural network module class.
+
+  A module is a named container for `tf.Variable`s, other `tf.Module`s and
+  functions which apply to user input. For example a dense layer in a neural
+  network might be implemented as a `tf.Module`:
+
+  >>> class Dense(tf.Module):
+  ...   def __init__(self, in_features, output_features):
+  ...     super(Linear, self).__init__()
+  ...     # NOTE: We're creating variables in the ctor, but you can actually
+  ...     #       create them in any method (including __call__) you like.
+  ...     self.w = tf.Variable(
+  ...         tf.random.normal([input_features, output_features]), name='w')
+  ...     self.b = tf.Variable(tf.zeros([output_features]), name='b')
+  ...
+  ...   def __call__(self, x):
+  ...     x = tf.convert_to_tensor(x, name='x')
+  ...     y = tf.matmul(x, self.w) + self.b
+  ...     return tf.nn.relu(y)
+
+  You can use the layer as you would expect:
+
+  >>> d = Dense(input_features=64, output_features=10)
+  >>> d(tf.ones([100, 64]))
+  <tf.Tensor: ...>
+
+  By subclassing `tf.Module` instead of `object` any variables created inside
+  the module are automatically created within the modules name scope:
+
+  >> d.w.name
+  "dense/w:0"
+
+  In eager mode this is useful for debugging, and when used with `@tf.function`
+  the use of name scopes gives operations (e.g. matmul) useful names as well.
+
+  As well as automatic naming, the Dense module inherits methods for tracking
+  its variables:
+
+  >>> d.variables
+  (<tf.Variable 'dense/b:0' ...>, <tf.Variable 'dense/w:0' ...>)
+  """
+
+  @property
+  def name(self):
+    """Returns the name of this module as passed or determined in the ctor.
+
+    NOTE: This is not the same as the `self.name_scope.name` which includes
+    parent module names.
+    """
+
+  @property
+  def name_scope(self):
+    """Returns a `tf.name_scope` instance for this class.
+
+    NOTE: The module name scope includes all parent modules as determined
+    when this module was created.
+    """
+
+  @property
+  def variables(self):
+    """Returns a tuple of variables owned by this module and it's submodules.
```

Comment:
Thanks for the IRL discussion, I'll use "Collection" in the docstring and still return a tuple at least for now.

Reply:
In general, returning a tuple makes the API a bit more efficient since you don't have to return a copy to avoid callers mutating your internal data structures.

Reply:
We agree returning a tuple is the right decision, but the signature/contract of the method should be less specific to allow this to change in future (e.g. if there is a faster immutable collection we might want to use that instead).

If we could include type hints in TF code I would go with the following:

```python
@property
def variables(self) -> typing.Collection[tf.Variable]:
  ...
```

Reply:
Perhaps `typing.Sequence` would be better here? Importantly, the returned collection has a well-defined, deterministic ordering.